### PR TITLE
[C++20][modules] Fix std::initializer_list recognition if it's exported out of a module

### LINKED ISSUE #118218: Clang: modules and std::initializer_list do not like each other
Clang version: 19.1.3
C++ version: 23
STL: from msvc

I was trying to compile nlohmann json using modules and I encountered an error (posting the original error for others who might google it):
```
C:\Users\username\.conan2\p\nlohm0567ffc90cfc1\p\include\nlohmann\detail\input\lexer.hpp:748:66: error: no matching constructor for initialization of 'std::initializer_list<char_int_type>' (aka 'initializer_list<int>')
  748 |                     if (JSON_HEDLEY_UNLIKELY(!next_byte_in_range({0x80, 0xBF})))
```

I tried to reduce code to the minimum reproducible example and here's what I got (pay attention to the highlighted lines changing which kind of fixes the issue):

<pre>
// main.cpp

import std; // <b><i>Workaround 1: if I delete this line everything works fine</i></b>
import mod;

int main() {
    lexer{}.func({1,1});
    return 0;
}
</pre>

```
// mod.cppm

module;

#include "inc.hpp"

export module mod;

export using ::lexer;
export using ::adapter;
```

<pre>
// inc.hpp

#pragma once

#include &lt;initializer_list>
#include &lt;string>

struct adapter {
    using char_type = char;
};

template&lt;typename InputAdapterType = adapter>
class lexer
{
    using char_type = typename InputAdapterType::char_type; // <b><i>Workaround 2: if I replace "InputAdapterType" here with "adapter" everything works fine</i></b>
    using char_int_type = typename std::char_traits&lt;char_type>::int_type;

public:
    void func(std::initializer_list&lt;char_int_type> ranges) {}
};
</pre>

<pre>
// std.cppm

module;

#define _BUILD_STD_MODULE // <b><i>Workaround 3: if I delete this line everything works fine</i></b>

#include &lt;intrin.h>

export module std;

#include &lt;initializer_list>
#include &lt;regex> // <b><i>Workaround 4: if I delete this line everything works fine</i></b>
</pre>

<details>
<summary>The full error I get</summary>

```
C:/Users/username/CLionProjects/untitled23_1/std/modules/std.msvc.cppm:9:10: warning: '#include <filename>' attaches the declarations to the named module 'std', which is not usually intended; consider moving that directive before the module declaration [-Winclude-angled-in-module-purview]
    9 | #include <initializer_list>
      |          ^
C:/Users/username/CLionProjects/untitled23_1/std/modules/std.msvc.cppm:7:15: warning: 'std' is a reserved name for a module [-Wreserved-module-identifier]
    7 | export module std;
      |               ^
C:/Users/username/CLionProjects/untitled23_1/std/modules/std.msvc.cppm:10:10: warning: '#include <filename>' attaches the declarations to the named module 'std', which is not usually intended; consider moving that directive before the module declaration [-Winclude-angled-in-module-purview]
   10 | #include <regex>
      |          ^
3 warnings generated.
[4/6] Building CXX object CMakeFiles/untitled23_1.dir/main.cpp.obj
FAILED: CMakeFiles/untitled23_1.dir/main.cpp.obj 
C:\Tools\llvm\bin\clang++.exe  -IC:/Users/username/CLionProjects/untitled23_1/include -IC:/Users/username/CLionProjects/untitled23_1/mod/../include -IC:/Users/username/CLionProjects/untitled23_1/std/include -O0 -std=c++23 -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -g -Xclang -gcodeview -MD -MT CMakeFiles/untitled23_1.dir/main.cpp.obj -MF CMakeFiles\untitled23_1.dir\main.cpp.obj.d @CMakeFiles\untitled23_1.dir\main.cpp.obj.modmap -o CMakeFiles/untitled23_1.dir/main.cpp.obj -c C:/Users/username/CLionProjects/untitled23_1/main.cpp
C:/Users/username/CLionProjects/untitled23_1/main.cpp:6:18: error: no matching constructor for initialization of 'std::initializer_list<char_int_type>' (aka 'initializer_list<int>')
    6 |     lexer{}.func({1,1});
      |                  ^~~~~
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.42.34433\include\initializer_list:33:15: note: candidate constructor not viable: no known conversion from 'int' to 'const int *' for 1st argument
   33 |     constexpr initializer_list(const _Elem* _First_arg, const _Elem* _Last_arg) noexcept
      |               ^                ~~~~~~~~~~~~~~~~~~~~~~~
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.42.34433\include\initializer_list:21:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
   21 | class initializer_list {
      |       ^~~~~~~~~~~~~~~~
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.42.34433\include\initializer_list:21:7: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 2 were provided
   21 | class initializer_list {
      |       ^~~~~~~~~~~~~~~~
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.42.34433\include\initializer_list:31:15: note: candidate constructor not viable: requires 0 arguments, but 2 were provided
   31 |     constexpr initializer_list() noexcept : _First(nullptr), _Last(nullptr) {}
      |               ^
C:\Users\username\CLionProjects\untitled23_1\mod\..\include\inc.hpp:17:52: note: passing argument to parameter 'ranges' here
   17 |     void func(std::initializer_list<char_int_type> ranges) {}
      |                                                    ^
1 error generated.
```

</details>

### PR DESCRIPTION
**Pull request by jijjijj**
If the std::initializer_list is exported out of module, its `DeclContext` is not a namespace as `Sema::isStdInitializerList` expects, but an `Decl::Kind::Export` and only its parent is a namespace. So this commit makes `Sema::isStdInitializerList` account for that.

I'm really new to clang so I'm not 100% sure that was the issue, it seems so and it fixes compilation. Also I probably need to add tests but I'd like someone to approve the idea first.

Fixes https://github.com/llvm/llvm-project/issues/118218

### CODE REVIEW

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 1]
```diff
@@ -11919,8 +11919,12 @@ bool Sema::isStdInitializerList(QualType Ty, QualType *Element) {
     CXXRecordDecl *TemplateClass = Template->getTemplatedDecl();
     if (TemplateClass->getIdentifier() !=
             &PP.getIdentifierTable().get("initializer_list") ||
-        !getStdNamespace()->InEnclosingNamespaceSetOf(
-            TemplateClass->getDeclContext()))
```
Comment (ChuanqiXu9) [inline clang/lib/Sema/SemaDeclCXX.cpp:11923]: Maybe we can fix it by replacing `getDeclContext()` with `getNonTransparentContext()`

**Review by ChuanqiXu9 — COMMENTED** [review]
The overall idea looks good. But the implementation may be improved. I didn't look at the implementation of std module in MSVC. But I am guess it is because the clang compiler failed to skip the transparent `export`  or `language linkage` context.

If yes, please attach a test and the release notes.

File: clang/test/Modules/initializer-list-recognition-through-export-and-linkage-issue-118218.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+//--- main.cpp
+// expected-no-diagnostics
+import std;
+import mod;
+
+int main() {
+  A{}.func({1,1});
+  return 0;
+}
```
Comment (ChuanqiXu9) [inline clang/test/Modules/initializer-list-recognition-through-export-and-linkage-issue-118218.cpp:39]: nit: new line

File: clang/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -803,6 +803,8 @@ Bug Fixes to C++ Support
 - Fixed an incorrect lambda scope of generic lambdas that caused Clang to crash when computing potential lambda
   captures at the end of a full expression. (#GH115931)
 - Clang no longer rejects deleting a pointer of incomplete enumeration type. (#GH99278)
+- Fixed recognition of ``std::initializer_list`` when it's surrounded with ``extern "C++"`` and exported
```
Comment (ChuanqiXu9) [inline clang/docs/ReleaseNotes.rst:806]: I guess it is worth mentioning the pattern comes from MSVC's std module implementation.

**Review by ChuanqiXu9 — COMMENTED** [review]
LGTM with nits

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM then

**Comment by ChuanqiXu9** [comment]
I'll merge this directly since it looks like you don't have commit access.

**Comment by ChuanqiXu9** [comment]
It looks nice to backport this if we're still in the release circle for 19.x

**Comment by ChuanqiXu9** [comment]
@jijjijj automated backport failed. You can create a PR to release/19.x manually to backport it.

**Comment by jijjijj** [comment]
@ChuanqiXu9 here's the PR: https://github.com/llvm/llvm-project/pull/121739

**Pull request merged**